### PR TITLE
GH-37993: [CI] Fix conda-integration build

### LINKED ISSUE #37993: [CI][Integration] conda-integration build failures
### Describe the bug, including details regarding any error messages, version, and platform.

The conda-integration build has recently started failing:
https://github.com/apache/arrow/actions/runs/6393852866/job/17353952453

Apparently this is because conda-forge is now providing Python 3.12 by default, and pythonnet [does not support it yet](https://github.com/pythonnet/pythonnet/pull/2249).

### Component(s)

Continuous Integration, Integration

### PR DESCRIPTION
**Pull request by pitrou**
### Rationale for this change

The conda-integration build has recently started failing:
https://github.com/apache/arrow/actions/runs/6393852866/job/17353952453

Apparently this is because conda-forge is now providing Python 3.12 by default, and pythonnet [does not support it yet](https://github.com/pythonnet/pythonnet/pull/2249).

### What changes are included in this PR?

Avoid using Python 3.12 for Archery in conda-integration build.

### Are these changes tested?

Yes, by construction.

### Are there any user-facing changes?

No.
* Closes: #37993

### CODE REVIEW

**Comment by pitrou** [comment]
Integration build was successful: https://github.com/apache/arrow/actions/runs/6393710784/job/17353516711?pr=37990

**Review by raulcd — APPROVED** [review]
LGTM, thanks @pitrou 
CI failures were unrelated I am re-running them but will merge once they finish.

**Comment by tustvold** [comment]
We are still seeing failures in arrow-rs - https://github.com/apache/arrow-rs/actions/runs/6405057279/job/17386849288

It appears to be using https://hub.docker.com/layers/apache/arrow-dev/amd64-conda-integration/images/sha256-8ed7ddbadbd9a6d32d66ab8896b7f05db34a42e34ba1f5d0ba4df2113dce6ce5?context=explore which was published after this was merged, so I would have expected it to contain this fix?

Any ideas?

**Comment by pitrou** [comment]
@tustvold Apparently you are calling `pip install -e dev/archery`. You instead need `pip install -e dev/archery[integration]`.

**Comment by pitrou** [comment]
Ah, and you must call `pip install pythonnet` explicitly as well.

**Pull request merged**
